rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,545 @@
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, hypothetical `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits (REFERENCE ONLY!)
+
+**These traits are provided as reference to illustrate how these primitive shape types might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.**
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+impl Meshable for Direction {}
+
+struct Plane {
+  point: Point,
+  normal: Direction,
+}
+impl Meshable for Plane {}
+
+/// Differentiates a line from a ray, where a line is infinite and a ray is directional half-line, although their underlying representation is the same.
+struct Ray(
+  point: Point, 
+  direction: Direction,
+);
+impl Meshable for Ray {}
+
+// Line types
+
+/// Unbounded line in 3D space with directionality
+struct Line { 
+  point: Point, 
+  direction: Direction,
+}
+impl Meshable for Line {}
+
```

Comment:
What exactly would the above three `Meshable` impls (for `Line`, `Ray`, and `Plane`) result in? They have infinite extent in some direction, so it's not feasible to generate a finite-size mesh.

Reply:
While the shapes are not finite, my thought was that a `Meshable` (or more likely the `From<Shape> for Mesh`) would output a mesh centered at the origin of the shapes. The goal for these was to have a very simple way of generating a mesh to debug these primitives.